modules/samples/firealarm/component/api/src/main/java/org/homeautomation/firealarm/api/DeviceTypeServiceImpl.java

```diff
@@ -0,0 +1,431 @@
+    private ConcurrentHashMap<String, DeviceJSON> deviceToIpMap = new ConcurrentHashMap<>();
+
+    private static String shortUUID() {
+        UUID uuid = UUID.randomUUID();
+        long l = ByteBuffer.wrap(uuid.toString().getBytes(StandardCharsets.UTF_8)).getLong();
+        return Long.toString(l, Character.MAX_RADIX);
+    }
+
+    /**
+     * @param agentInfo device owner,id
+     * @return true if device instance is added to map
+     */
+    @Path("device/register")
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response registerDevice(final DeviceJSON agentInfo) {
+        String deviceId = agentInfo.deviceId;
+        if ((agentInfo.deviceId != null) && (agentInfo.owner != null)) {
+            deviceToIpMap.put(deviceId, agentInfo);
+            return Response.status(Response.Status.OK).build();
+        }
+        return Response.status(Response.Status.NOT_ACCEPTABLE).build();
+    }
+
+    /**
+     * @param deviceId unique identifier for given device type instance
+     * @param state    change status of sensor: on/off
+     */
+    @Path("device/{deviceId}/change-status")
+    @POST
+    @Feature(code = "change-status", name = "Change status of sensor: on/off",
+            description = "Change status of sensor: on/off")
+    public Response changeStatus(@PathParam("deviceId") String deviceId,
+                                 @QueryParam("state") String state,
+                                 @Context HttpServletResponse response) {
+        try {
+            if (!APIUtil.getDeviceAccessAuthorizationService().isUserAuthorized(new DeviceIdentifier(deviceId,
+                                                                                                     DeviceTypeConstants.DEVICE_TYPE))) {
+                return Response.status(Response.Status.UNAUTHORIZED.getStatusCode()).build();
+            }
+            String sensorState = state.toUpperCase();
+            if (!sensorState.equals(DeviceTypeConstants.STATE_ON) && !sensorState.equals(
+                    DeviceTypeConstants.STATE_OFF)) {
+                log.error("The requested state change should be either - 'ON' or 'OFF'");
+                return Response.status(Response.Status.BAD_REQUEST.getStatusCode()).build();
+            }
+            Map<String, String> dynamicProperties = new HashMap<>();
+            String publishTopic = APIUtil.getAuthenticatedUserTenantDomain()
+                                  + "/" + DeviceTypeConstants.DEVICE_TYPE + "/" + deviceId + "/command";
+            dynamicProperties.put(DeviceTypeConstants.ADAPTER_TOPIC_PROPERTY, publishTopic);
+            APIUtil.getOutputEventAdapterService().publish(DeviceTypeConstants.MQTT_ADAPTER_NAME,
+                                                           dynamicProperties, state);
+            return Response.ok().build();
+        } catch (DeviceAccessAuthorizationException e) {
+            log.error(e.getErrorMessage(), e);
```

Comment:
Better to use an error message specific to API layer rather than the error message that can be retrieved from exception object, which is already included in the printed stack-trace.